### ISSUE TITLE
Main function not reached

### ISSUE THREAD

**Issue by patroclo77** [AdaCore/RC_Car_Demo#3]
Hi,
after compiling and testing everything I have realised that the main function is not reached. In other words, in the generated file: b__rc_car.adb:

begin
      Initialize (SEH'Address);
      adainit;
      Ada_Main_Program; // Code not reached.
      adafinal;
      Finalize;

Find the object code and firmware attached. Tested both in QEMU and on-target.

Many thanks in advance
[Production.zip](https://github.com/AdaCore/RC_Car_Demo/files/4332491/Production.zip)

**Comment by pat-rogers** [comment]
On 14-Mar-20 4:51 AM, patroclo77 wrote:
> Hi,
> after compiling and testing everything I have realised that the main 
> function is not reached. In other words, in the generated file: 
> b__rc_car.adb:
> 
> begin
> Initialize (SEH'Address);
> adainit;
> Ada_Main_Program; // Code not reached.
> adafinal;
> Finalize;

This almost always indicates that the elaboration of a library package 
has raised an exception.

The underlying issue is dependencies on the actual hardware platform, ie 
the car and powerplant.

For example, if you debug the app on the emulator, you will see that the 
  Remote_Control package's procedure Initialize will raise an exception 
regarding the IR receiver's I2C clock frequency. The requested frequency 
  is possible on the actual board's I2C device but not on the emulator.

Even after the main procedure runs, the code still has a dependency on 
the actual hardware. For example, the Steering_Control package's Servo 
task initializes the steering mechanism to find the steering offset for 
the "zero" angle. At the end of the procedure, the offset must be 
greater than zero if the initialization worked at all, and so we have a 
pragma Assume to that effect for the sake of proving the postcondition. 
We must have power to the steering motor for that to work, otherwise the 
check for pragma Assume fails when run-time checks are enabled.

So bottom line, to really debug it you will need the actual STM board, 
or one compatible, and also either build the car and apply power to the 
steering motor, or just comment-out the pragma Apply in procedure 
Initialize_Steering_Mechanism.

There may be other cases as well. As you can imagine, I debug directly 
on the RC car and its STM32F4 board so I can turn on the battery before 
telling the debugger to "go".

Good luck, and feel free to ask if you have any further questions about this.

**Comment by patroclo77** [comment]
Hi,
Thanks for the help. I do have the board but not peripherals. Yet it launches some form of exception when it boots. I also tried the "Debug" configuration but still no luck.

Best regards.

**Comment by pat-rogers** [comment]
On 16-Mar-20 2:22 PM, patroclo77 wrote:
> Hi,
> Thanks for the help. I do have the board but not peripherals. Yet it 
> launches some form of exception when it boots. I also tried the "Debug" 
> configuration but still no luck.

Yes, I believe it is raising an exception from procedure 
Steering_Control.Initialize for lack of a motor and power.  If you 
comment out the pragma Assume on line 200 of that package it should be OK.

Overall, you should enable stopping on any exception. I never use 
exceptions for flow control so they are not expected and indicate a real 
problem.

To do so, once in the debugger you will be in the Debug perspective in 
GNATstudio, so you can open the Breakpoints view via the 
Debug->Data->Breakpoints menu.

In that view, press the "+" sign icon to create a new breakpoint. In the 
resulting dialog box, use the pull-down menu at the top to select "break 
on exception" and then make sure that "All Ada exceptions" is selected, 
as well as "Stop always." The rest should be empty/disabled by default. 
Press the OK button. You should now see this "catch-all" breakpoint in 
the list of all breakpoints in that view.

Now, when you run the program in the debugger, it should stop on any 
line that raises any exception. You can then use the Call Stack view to 
trace the sequence of calls leading to the exception.
